app/src/main/kotlin/compiler/intermediate_form/Pattern.kt

```diff
@@ -0,0 +1,166 @@
+package compiler.intermediate_form
+
+import kotlin.reflect.KClass
+import kotlin.reflect.cast
+
+sealed class Pattern {
+
+    // Returns null if the pattern doesn't match provided intermediate form tree
+    // If the tree matches, it returns the list of unmatched subtrees and
+    // captured values (mainly used to store values in matched leaves)
+    abstract fun match(node: IntermediateFormTreeNode): Pair<List<IntermediateFormTreeNode>, Map<String, Any>>?
+
+    sealed class ArgumentPattern<T> {
+        abstract fun match(value: T): Map<String, Any>?
+    }
+
+    // Matches any value of type T and maps it to the provided name
+    data class AnyArgument<T>(val name: String? = null) : ArgumentPattern<T>() {
+        override fun match(value: T): Map<String, Any>? {
+            return if (name == null) emptyMap() else mapOf(name to value as Any)
+        }
+    }
+
+    // Matches any value in allowed set and maps it to the provided name
+    data class ArgumentOf<T>(private val allowed: Set<T>, val name: String? = null) : ArgumentPattern<T>() {
+        override fun match(value: T): Map<String, Any>? {
+            if (!allowed.contains(value)) return null
+            return if (name == null) emptyMap() else mapOf(name to value as Any)
+        }
+    }
+
+    data class ArgumentWhere<T>(val name: String? = null, val predicate: (T) -> Boolean) : ArgumentPattern<T>() {
+        override fun match(value: T): Map<String, Any>? {
+            if (!predicate(value)) return null
+            return if (name == null) emptyMap() else mapOf(name to value as Any)
+        }
+    }
+
+    // Matches every intermediate form tree node
+    class AnyNode : Pattern() {
+        override fun match(node: IntermediateFormTreeNode): Pair<List<IntermediateFormTreeNode>, Map<String, Any>>? {
+            return Pair(listOf(node), emptyMap())
+        }
+    }
+
+    // Matches the first matching pattern in the provided list
+    // Stores the name of the pattern under name
+    data class FirstOf(val name: String, val possiblePatterns: List<Pair<Any, Pattern>>) : Pattern() {
+        override fun match(node: IntermediateFormTreeNode): Pair<List<IntermediateFormTreeNode>, Map<String, Any>>? {
+            for ((patternName, pattern) in possiblePatterns) {
+                val match = pattern.match(node)
+                if (match != null) {
+                    return Pair(match.first, match.second + mapOf(name to patternName))
+                }
+            }
+            return null
+        }
+    }
+
+    data class BinaryOperator<T : IntermediateFormTreeNode.BinaryOperator>(
+        val type: KClass<T>,
+        val leftPattern: Pattern = AnyNode(),
+        val rightPattern: Pattern = AnyNode()
+    ) : Pattern() {
+        override fun match(node: IntermediateFormTreeNode): Pair<List<IntermediateFormTreeNode>, Map<String, Any>>? {
+            if (!type.isInstance(node)) return null
```

Comment:
How about:
`val castedNode = type.safeCast(node) ?: return null`
and then using castedNode. Less reflection in the code...
I consider it cleaner, but I guess it's not evidently better.